homeassistant/helpers/llm.py

```diff
@@ -0,0 +1,240 @@
+"""Module to coordinate llm tools."""
+
+from __future__ import annotations
+
+from abc import abstractmethod
+from collections.abc import Callable, Iterable
+from dataclasses import dataclass
+from functools import cached_property
+import logging
+from types import FunctionType
+from typing import Any
+
+import voluptuous as vol
+from voluptuous_openapi import UNSUPPORTED, convert
+
+from homeassistant.core import Context, HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
+
+from . import area_registry as ar, config_validation as cv, floor_registry as fr, intent
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_KEY = "llm_tool"
+
+
+def default_custom_serializer(schema: Any) -> Any:
+    """Serialize additional types in OpenAPI-compatible format."""
+    from homeassistant.util.color import (  # pylint: disable=import-outside-toplevel
+        color_name_to_rgb,
+    )
+
+    if schema is cv.string:
+        return {"type": "string"}
+
+    if schema is cv.boolean:
+        return {"type": "boolean"}
+
+    if schema is color_name_to_rgb:
+        return {"type": "string"}
+
+    if isinstance(schema, cv.multi_select):
+        return {"enum": schema.options}
+
+    if isinstance(schema, FunctionType):
+        return {}
+
+    return UNSUPPORTED
+
+
+@dataclass(slots=True)
+class ToolInput:
+    """Tool input to be processed."""
+
+    tool_name: str
+    tool_args: dict[str, Any]
+    platform: str
+    context: Context | None
+    user_prompt: str | None
+    language: str | None
+    agent_id: str | None
+    conversation_id: str | None
+    device_id: str | None
+    assistant: str | None
+
+
+class Tool:
+    """LLM Tool base class."""
+
+    name: str
+    description: str
+    parameters: vol.Schema
+    custom_serializer: Callable[[Any], Any]
+
+    def __init__(
+        self,
+        name: str,
+        description: str,
+        parameters: vol.Schema = vol.Schema({}),
+        custom_serializer: Callable[[Any], Any] = default_custom_serializer,
+    ) -> None:
```

Comment:
Given some of the other comments, I am wondering if this constructor can be dropped. When I was talking offline about not having the tool object I think the thing I didn't like was the inheritance and constructor arguments.  But maybe if the style is that these are all class attributes It might be easier for the caller to implement e.g.:
```
class MyTool(llm.Tool):
    """This is my tool."""

    name = "my_tool"
    description = "My tool is great"
    schema = vol.Schema({... })

    def async_handle(...)
```

If it works this way then I think I can see why this is more desirable than passing a bunch of  arguments to the registration method.  WDYT?

Reply:
As I understand it:
If you are creating the class, then redefining the attributes is easier.
If you are creating the object, however, then the constructor is easier.
The former works well when one class usually represents a single tool. The latter works well when one class represents multiple tools, such as `IntentTool` objects can represent several intents.
I can remove the constructor in the base Tool class and replace the constructor with `@classmethod` for the `IntentTool` if you think that is better.

Reply:
It would require a `staticmethod` decorator:
```python
class Tool:
    ...
    custom_serializer: Callable[[Any], Any] = staticmethod(default_custom_serializer)
```

Reply:
Done that.